File: AIReader/AIReader/Support/ZIPArchive.swift
import Foundation

/// Minimal read-only ZIP reader: enough of the format to unpack an EPUB.
///
/// Only the stored and deflated methods are supported, which is everything an
/// EPUB is allowed to use.
struct ZIPArchive {
    enum ReadError: LocalizedError {
        case notAZIPArchive
        case unsupportedCompression(UInt16)
        case corruptEntry(String)

        var errorDescription: String? {
            switch self {
            case .notAZIPArchive: "The file is not a ZIP archive."
            case let .unsupportedCompression(method): "Unsupported ZIP compression method \(method)."
            case let .corruptEntry(path): "The archive entry “\(path)” could not be read."
            }
        }
    }

    struct Entry {
        let path: String
        let compressionMethod: UInt16
        let compressedSize: Int
        let headerOffset: Int
    }

    private let data: Data
    private(set) var entries: [String: Entry] = [:]

    init(data: Data) throws {
        self.data = data
        guard let directory = Self.locateCentralDirectory(in: data) else {
            throw ReadError.notAZIPArchive
        }

        var offset = directory.offset
        for _ in 0..<directory.count {
            guard offset + 46 <= data.count, u32(data, offset) == 0x0201_4b50 else { break }
            let nameLength = Int(u16(data, offset + 28))
            let extraLength = Int(u16(data, offset + 30))
            let commentLength = Int(u16(data, offset + 32))
            let path = string(data, offset + 46, nameLength)
            entries[path] = Entry(
                path: path,
                compressionMethod: u16(data, offset + 10),
                compressedSize: Int(u32(data, offset + 20)),
                headerOffset: Int(u32(data, offset + 42))
            )
            offset += 46 + nameLength + extraLength + commentLength
        }
    }

    /// Decompresses one entry. Paths are the archive-internal paths.
    func contents(of path: String) throws -> Data {
        guard let entry = entries[path] else { throw ReadError.corruptEntry(path) }
        let header = entry.headerOffset
        guard header + 30 <= data.count, u32(data, header) == 0x0403_4b50 else {
            throw ReadError.corruptEntry(path)
        }

        // The local header repeats the name and extra field lengths, which are
        // allowed to differ from the central directory's.
        let start = header + 30 + Int(u16(data, header + 26)) + Int(u16(data, header + 28))
        let end = start + entry.compressedSize
        guard end <= data.count else { throw ReadError.corruptEntry(path) }
        let payload = data.subdata(in: (data.startIndex + start)..<(data.startIndex + end))

        switch entry.compressionMethod {
        case 0:
            return payload
        case 8:
            guard let inflated = Inflate.raw(payload) else { throw ReadError.corruptEntry(path) }
            return inflated
        default:
            throw ReadError.unsupportedCompression(entry.compressionMethod)
        }
    }

    // MARK: - Byte reading

    /// Scans backwards for the end-of-central-directory record, which sits at
    /// the very end of the file followed only by an optional comment.
    private static func locateCentralDirectory(in data: Data) -> (offset: Int, count: Int)? {
        let minimum = 22
        guard data.count >= minimum else { return nil }
        let earliest = max(0, data.count - minimum - 0xFFFF)
        var offset = data.count - minimum
        while offset >= earliest {
            if u32(data, offset) == 0x0605_4b50 {
                return (Int(u32(data, offset + 16)), Int(u16(data, offset + 10)))
            }
            offset -= 1
        }
        return nil
    }

    private static func u16(_ data: Data, _ offset: Int) -> UInt16 {
        let base = data.startIndex + offset
        return UInt16(data[base]) | UInt16(data[base + 1]) << 8
    }

    private static func u32(_ data: Data, _ offset: Int) -> UInt32 {
        UInt32(u16(data, offset)) | UInt32(u16(data, offset + 2)) << 16
    }

    private func u16(_ data: Data, _ offset: Int) -> UInt16 { Self.u16(data, offset) }
    private func u32(_ data: Data, _ offset: Int) -> UInt32 { Self.u32(data, offset) }

    private func string(_ data: Data, _ offset: Int, _ length: Int) -> String {
        let base = data.startIndex + offset
        let bytes = data.subdata(in: base..<(base + length))
        return String(data: bytes, encoding: .utf8) ?? ""
    }
}

File: AIReaderKindle/src/Support/ZipArchive.cpp
#include "ZipArchive.hpp"

#include "Files.hpp"
#include "Inflate.hpp"
#include "Text.hpp"

std::optional<ZipArchive> ZipArchive::open(const std::string& path, std::string* error) {
    ZipArchive archive;
    auto data = Files::read(path);
    if (!data) {
        if (error) *error = "The file could not be read.";
        return std::nullopt;
    }
    archive.data_ = std::move(*data);
    if (!archive.readCentralDirectory()) {
        if (error) *error = "The file is not a ZIP archive.";
        return std::nullopt;
    }
    return archive;
}

std::vector<std::string> ZipArchive::paths() const {
    std::vector<std::string> result;
    for (const auto& entry : entries_) result.push_back(entry.first);
    return result;
}

std::optional<std::string> ZipArchive::find(const std::string& path) const {
    if (entries_.count(path)) return path;
    std::string wanted = Text::lower(path);
    for (const auto& entry : entries_) {
        if (Text::lower(entry.first) == wanted) return entry.first;
    }
    return std::nullopt;
}

/// Scans backwards for the end-of-central-directory record, which sits at the
/// very end of the file followed only by an optional comment.
bool ZipArchive::readCentralDirectory() {
    const size_t minimum = 22;
    if (data_.size() < minimum) return false;
    size_t earliest = data_.size() > minimum + 0xFFFF ? data_.size() - minimum - 0xFFFF : 0;
    size_t offset = data_.size() - minimum;
    while (true) {
        if (u32(offset) == 0x06054b50) break;
        if (offset == earliest) return false;
        --offset;
    }

    size_t count = u16(offset + 10);
    size_t position = u32(offset + 16);
    for (size_t i = 0; i < count; ++i) {
        if (position + 46 > data_.size() || u32(position) != 0x02014b50) break;
        size_t nameLength = u16(position + 28);
        size_t extraLength = u16(position + 30);
        size_t commentLength = u16(position + 32);
        if (position + 46 + nameLength > data_.size()) break;
        std::string path = data_.substr(position + 46, nameLength);
        entries_[path] = Entry{u16(position + 10), u32(position + 20), u32(position + 42)};
        position += 46 + nameLength + extraLength + commentLength;
    }
    return true;
}

std::optional<std::string> ZipArchive::contents(const std::string& path) const {
    auto found = entries_.find(path);
    if (found == entries_.end()) return std::nullopt;
    const Entry& entry = found->second;
    size_t header = entry.headerOffset;
    if (header + 30 > data_.size() || u32(header) != 0x04034b50) return std::nullopt;

    // The local header repeats the name and extra field lengths, which are
    // allowed to differ from the central directory's.
    size_t start = header + 30 + u16(header + 26) + u16(header + 28);
    size_t end = start + entry.compressedSize;
    if (end > data_.size()) return std::nullopt;
    std::string payload = data_.substr(start, entry.compressedSize);

    switch (entry.method) {
    case 0: return payload;
    case 8: return Inflate::raw(payload);
    default: return std::nullopt;
    }
}

uint16_t ZipArchive::u16(size_t offset) const {
    return static_cast<uint8_t>(data_[offset]) | static_cast<uint8_t>(data_[offset + 1]) << 8;
}

uint32_t ZipArchive::u32(size_t offset) const {
    return static_cast<uint32_t>(u16(offset)) | static_cast<uint32_t>(u16(offset + 2)) << 16;
}
